components/playlist/browser/playlist_service.cc

```diff
@@ -424,19 +446,27 @@ void PlaylistService::FindMediaFilesFromContents(
     content::WebContents* contents,
     FindMediaFilesFromContentsCallback callback) {
   CHECK(*enabled_pref_) << "Playlist pref must be enabled";
-  DCHECK(contents);
+  CHECK(contents);
 
-  PlaylistDownloadRequestManager::Request request;
   auto current_url = contents->GetVisibleURL();
-  if (ShouldGetMediaFromBackgroundWebContents(contents)) {
-    request.url_or_contents = current_url.spec();
-  } else {
-    request.url_or_contents = contents->GetWeakPtr();
+  auto* tab_helper = PlaylistTabHelper::FromWebContents(contents);
+  CHECK(tab_helper);
+
+  for (const auto& item : tab_helper->found_items()) {
+    if (download_request_manager_->ShouldRefetchMediaSourceToCache(item)) {
+      PlaylistDownloadRequestManager::Request request;
+      request.url = current_url;
+      request.should_force_fake_ua = ShouldUseFakeUA(current_url);
+      request.callback = base::BindOnce(std::move(callback), current_url);
+      download_request_manager_->GetMediaFilesFromPage(std::move(request));
+      return;
+    }
   }
 
-  request.should_force_fake_ua = ShouldUseFakeUA(contents->GetVisibleURL());
-  request.callback = base::BindOnce(std::move(callback), current_url);
-  download_request_manager_->GetMediaFilesFromPage(std::move(request));
+  std::vector<mojom::PlaylistItemPtr> items;
+  base::ranges::transform(tab_helper->found_items(), std::back_inserter(items),
+                          &mojom::PlaylistItemPtr::Clone);
+  std::move(callback).Run(current_url, std::move(items));
```

Comment:
Nit — `base::ranges::none_of<>()` might make it a bit more readable:
```cpp
if (base::ranges::none_of(tab_helper->found_items(), [&](const auto& item) {
      return download_request_manager_->ShouldRefetchMediaSourceToCache(item);
    })) {
  std::vector<mojom::PlaylistItemPtr> items;
  base::ranges::transform(tab_helper->found_items(),
                          std::back_inserter(items),
                          &mojom::PlaylistItemPtr::Clone);
  return std::move(callback).Run(current_url, std::move(items));
}

PlaylistDownloadRequestManager::Request request;
request.url = current_url;
request.should_force_fake_ua = ShouldUseFakeUA(current_url);
request.callback = base::BindOnce(std::move(callback), current_url);
download_request_manager_->GetMediaFilesFromPage(std::move(request));
```

Reply:
Also, it's unfortunate that we have to do this check here as well, as one of the paths ends up calling `PlaylistService::FindMediaFilesFromContents()` through:
- `AddBubble::AddBubble()`
- `PlaylistTabHelper::ShouldRefetch()`
- `PlaylistService::ShouldRefetchMediaSourceToCache()`
- `PlaylistDownloadRequestManager::ShouldRefetchMediaSourceToCache()`
- ...
- `PlaylistTabHelper::RefetchMediaURL()`
- `PlaylistTabHelper::FindMediaFromCurrentContents()`
- `PlaylistService::FindMediaFilesFromContents()`

which already has this check (`PlaylistDownloadRequestManager::ShouldRefetchMediaSourceToCache()`).

Reply:
> Also, it's unfortunate that we have to do this check here as well, as one of the paths ends up calling PlaylistService::FindMediaFilesFromContents() through:

Yes, that's true. I put the check considering the Android case, which calls FindMediaFilesFromActiveTab.